### ISSUE TITLE
`torch.compile` cannot compile a model with a basic `LSTM`, even on latest main

### ISSUE THREAD

**Issue by tinywisdom** [pytorch/pytorch#158007]
## Reproduction Script
```python
import torch
import torch.nn as nn

class MyModel(nn.Module):
    def __init__(self, rnn_dims=512):
        super().__init__()
        self.rnn = nn.LSTM(input_size=rnn_dims, hidden_size=rnn_dims, batch_first=True)

    def forward(self, x):
        h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
        c0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
        out, h1 = self.rnn(x, (h0, c0))
        return h1  # tuple (h_n, c_n)

if __name__ == "__main__":
    torch.manual_seed(0)
    model = MyModel().eval().to("cuda")
    input_tensor = torch.rand(1, 1, 512, dtype=torch.float32).to("cuda")

    try:
        with torch.no_grad():
            eager_output = model(input_tensor)
        print("Eager output:", eager_output[0].shape, eager_output[1].shape)
    except Exception as e:
        print(e)

    try:
        compiled_model = torch.compile(model, fullgraph=True, backend="inductor")
        with torch.no_grad():
            compiled_output = compiled_model(input_tensor)
        print("Compiled output:", compiled_output[0].shape, compiled_output[1].shape)
    except Exception as e:
        print(e)
```

## Output on Stable (2.7.1) and Latest `main` (2.9.0a0+gitfcc682b)

<details> <summary>Output (Stable)</summary>

```
Eager output: torch.Size([1, 1, 512]) torch.Size([1, 1, 512])
TorchDynamo purposely graph breaks on RNN, GRU, LSTMs

from user code:
   File "for_test.py", line 10, in forward
    h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
```

</details> 

<details> <summary>Output (main branch)</summary>

```
Eager output: torch.Size([1, 1, 512]) torch.Size([1, 1, 512])
Attempted to wrap RNN, GRU, or LSTM
  Explanation: Dynamo does not support RNN, GRU, or LSTM.
  Hint: It may be possible to write Dynamo tracing rules for this code. Please report an issue to PyTorch if you encounter this graph break often and it is causing performance issues.

  Developer debug context: LSTM(512, 512, batch_first=True)

from user code:
   File "test_coverage/for_test.py", line 10, in forward
    h0 = torch.zeros(1, x.size(0), self.rnn.hidden_size, device=x.device, dtype=x.dtype)
```

</details>

## What’s Confusing?
+ This model runs just fine under eager execution.

+ It’s a very basic LSTM-based model, and doesn’t use anything exotic or fused.

+ Yet torch.compile breaks, even on the latest main (torch==2.9.0a0+gitfcc682b).

+ I understand some ops are not supported yet — but shouldn't a warning or fallback happen, instead of an outright failure? Or is there some plan to support nn.LSTM and friends via fake tensor rules in the future?

### Versions

Both `torch==2.9.0a0+gitfcc682b` and `torch==2.7.1a0+gitfcc682b` have the same problem.

```
PyTorch version: 2.7.1a0+gite2d141d
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

Thanks for the great work on `torch.compile` — I appreciate any clarification on this!

**Comment by theap06** [comment]
Can I take this issue up?

**Comment by xmfan** [comment]
> I understand some ops are not supported yet — but shouldn't a warning or fallback happen, instead of an outright failure?

The fallback is to graph break, but specifying `fullgraph=True` will force a loud error
```python
    try:
        compiled_model = torch.compile(model, fullgraph=True, backend="inductor") 
        with torch.no_grad():
```

**Comment by XAheli** [comment]
Investigated on latest `main` (`2.13.0a0+git795be92`, CUDA H200) with my agent. Sharing runtime-traced findings that recharacterize the problem.

### Finding 1: `allow_rnn=True` works. dynamo traces LSTM/GRU successfully

With `torch._dynamo.config.allow_rnn = True`, dynamo bypasses the `builder.py` isinstance block and **successfully traces through** LSTM and GRU forward methods. The resulting FX graph keeps `aten.lstm` as a single op:

```
1-layer LSTM (h=5, seq=3):   12 FX nodes → compiled in 0.7s   [output matches eager (atol=1e-4)]
1-layer GRU  (h=5, seq=3):   compiled in 1.0s         [output matches eager (atol=1e-3)]
```

The compiled model's outputs match eager within normal float32 precision (~1e-4), and recompilation produces **exact bitwise matches**, confirming determinism.

### Finding 2: The reported "hang" is not in dynamo it's in inductor's `compile_fx`

For larger models, the issue isn't dynamo failing to trace — it's the inductor backend taking extremely long to lower the `aten.lstm` op. The FX graph stays small even for 6-layer LSTMs:

| Config | FX nodes | `compile_fx` time | Status |
|--------|----------|-------------------|--------|
| 1L, h=5, seq=3 | 12 | 0.7s | pass |
| 1L, h=64, seq=32 | 12 | 18.8s | pass |
| 2L, h=64, seq=16 | 16 | 5.6s | pass |
| 6L, h=256, seq=64 | **32** | **79.9s** | barely |
| 1L, h=256, seq=128 | 12 | >60s | timeout |
| 6L, h=256, seq=512 | 32 | >120s | timeout |

The FX graph for a 6-layer LSTM is only **32 nodes** (25 placeholders, 6 call_functions, 1 output), yet inductor's `compile_fx` spends 80 seconds. Timeout tracebacks consistently land in inductor subsystems: `scheduler.py:_can_fuse`, `pattern_matcher.py:_match`, and `graph.py:get_dep_size_hint`.

This is a separate inductor performance issue — the `aten.lstm` op's decomposition/lowering inside inductor produces a much larger internal graph that these passes struggle with.

### Finding 3: Without `fullgraph=True`, graph break recovery is broken

Related to #141184: when `allow_rnn=False` (default) and `fullgraph` is not set, dynamo graph-breaks on the LSTM but then **fails to compile any subgraph at all** — neither pre-LSTM nor post-LSTM. Dynamo emits:

```
empty checkpoint - cannot resume from graph break
Failed to handle graph break gracefully. Skipping the function and falling back to eager.
```

Result: 0 compiled graphs, the entire model runs in eager.

## Summary

There are actually **three distinct issues** here:
1. **Dynamo blocks RNN by default** (`allow_rnn=False`) — intentional, but the `allow_rnn=True` escape hatch works
2. **Graph break recovery is broken** (0 graphs compiled) — tracked at #141184
3. **Inductor `compile_fx` scales poorly** for `aten.lstm` lowering — 32-node FX graph takes 80s, making multi-layer/large LSTMs practically unusable. This seems to be the actual blocker and may warrant a separate inductor issue.

## Environment

```
PyTorch: 2.13.0a0+git795be92
CUDA: 13.0
GPU: NVIDIA H200
OS: RHEL 9 (kernel 5.14.0-615.el9.x86_64)
```
